### PR TITLE
chore: Set default return type to ExportedProgram

### PR DESCRIPTION
**Pull request by peri044**
# Description

This change adds the following

1) Returns ExportedProgram as the output of compilation by default
2) `output_format` flag to return desired output format (exported_program | torchscript | graph_module)
3) Modify export serde test cases to handle the above changes

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 1]
```diff
@@ -45,19 +43,11 @@ def export(
                         format=torch.channel_last
                     ), # Dynamic input shape for input #2
                     torch.randn((1, 3, 224, 244)) # Use an example tensor and let torch_tensorrt infer settings
-        ir (str): torchscript | exported_program. Based on the provided ir, the output type would be a torchscript or exported program.
     """
-    if ir == "torchscript":
-        return torch.jit.trace(gm, inputs)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:51]: What are the options for users who want to `torchscript` the exported product for serialization?

**Review by gs-olive — COMMENTED** [review]
Overall looks good. Added a question, and it seems 2 tests are failing currently. Consider redirecting the branch to the `staging_rel_2_2` branch to get the release test suite.

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 1 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/_exporter.py:51]: They could always post return trace

Comment (narendasan) [reply py/torch_tensorrt/dynamo/_exporter.py:51]: Alternatively, we can have a string/enum, input for `output_format` with options: `exported_program/ep`, `torchscript/ts`, `graph_module/fx`

**Review by narendasan — COMMENTED** [review]
We need documentation updates in this PR

**Comment by peri044** [comment]
This PR has an open issue https://github.com/pytorch/pytorch/issues/116831 which should be fixed first

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 1 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py]: ^ This looks good. Added `output_format` now

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 2]
```diff
@@ -84,34 +72,78 @@ def transform(
...
+
+            # state_dict has these parameters/buffers as torch.Tensors. We override them as torch.nn.Parameter/torch.Tensors respectively.
+            for name, _ in gm.named_parameters():
+                if node.target == name:
+                    input_kind = InputKind.PARAMETER
+                    state_dict[name] = constant_tensor
+            for name, _ in gm.named_buffers():
+                if node.target == name:
+                    input_kind = InputKind.BUFFER
+                    state_dict[name] = constant_tensor
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:118]: Consider using `break` once the name is matched to a parameter or buffer. Are the `get_parameter`/`get_buffer` functions useful for this case?

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 3]
```diff
@@ -84,34 +72,78 @@ def transform(
...
+    # The input_specs should be of the form [params, buffers, constant_tensors, user_inputs]
+    non_user_input_idx = 0
+    for node in gm.graph.nodes:
         if node.op == "get_attr":
-            constant_tensor = getattr(trt_gm, node.target)
-            with trt_gm.graph.inserting_before(first_user_input):
-                const_placeholder_node = trt_gm.graph.placeholder(node.target)
-                const_placeholder_node.meta = copy.deepcopy(node.meta)
+            constant_tensor = getattr(gm, node.target)
+            input_kind = InputKind.CONSTANT_TENSOR
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:108]: In the constant tensor case, does the `state_dict` not need to be updated?

File: py/torch_tensorrt/dynamo/_exporter.py [hunk 4]
```diff
@@ -84,34 +72,78 @@ def transform(
...
                     constant_tensor
                 )
                 node.replace_all_uses_with(const_placeholder_node)
-                trt_gm.graph.erase_node(node)
+                gm.graph.erase_node(node)
+
+                # Add these parameters/buffers/constants to the existing graph signature
+                # before user inputs. These specs are looked up in the state_dict during ExportedProgram creation.
+                graph_signature.input_specs.insert(
+                    non_user_input_idx,
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/_exporter.py:134]: Is it critical that the `[params, buffers, constant_tensors, user_inputs]` ordering above be strictly followed? It seems that in certain suboptimal node interleavings, params can come after buffers if the `get_attr` nodes are ordered differently.

**Review by gs-olive — COMMENTED** [review]
Overall, looks great - just added a few questions on the exporter.

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 2 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py:118]: Added a break. `get_parameter` apparently raises an `AttributeError`https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.get_parameter  if it's not found. We could wrap it in a try-except but I'll leave it like this for now unless there's a strong preference. All we want here is if the particular node is a parameter, assign `input_kind` appropriately and override the state_dict

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 3 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py:108]: Yeah no need because they are of the type `torch.Tensor` which is correct

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 4 above)
Comment (peri044) [reply py/torch_tensorrt/dynamo/_exporter.py:134]: `[params, buffers, constant_tensors, user_inputs]` is the convention that I saw being followed within Pytorch code. From my understanding, the ordering of input specs should follow the order of `get_attr` nodes which is what the code does.

File: py/torch_tensorrt/dynamo/_exporter.py (hunk 2 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/_exporter.py]: This sounds good to me - nested `try/except` sounds pretty messy
